File: LEETCODE/LEETCODE-JAVA/1395.java
//1395. Count Number of Teams
//https://leetcode.com/problems/count-number-of-teams/

class Solution {
    public int numTeams(int[] rating) {
        int n=rating.length;

        int maxAdjSize=0;
        for(int val: rating){
            maxAdjSize=Math.max(maxAdjSize,val);
        }

        List<List<Integer>> adj1=new ArrayList<>();
        List<List<Integer>> adj2=new ArrayList<>();

        for(int i=0;i<=maxAdjSize;i++){
            adj1.add(new ArrayList<>());
            adj2.add(new ArrayList<>());
        }

        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                if(rating[i]>rating[j]){
                    adj1.get(rating[i]).add(rating[j]);
                }
                else if(rating[i]<rating[j]){
                    adj2.get(rating[i]).add(rating[j]);
                }
            }
        }

        int res=0;
        for(int val:rating){
            for(int next:adj1.get(val)){
                res+=adj1.get(next).size();
            }

            for(int next:adj2.get(val)){
                res+=adj2.get(next).size();
            }
        }

        return res;
    }
}

File: LEETCODE/LEETCODE-CPP/1395.cpp
// 1395. Count Number of Teams
// https://leetcode.com/problems/count-number-of-teams/

// Method-1 using adj matrices
class Solution
{
public:
    int numTeams(vector<int> &rating)
    {
        int n = rating.size();
        int maxAdjSize = 0;
        for (int &i : rating)
        {
            maxAdjSize = max(maxAdjSize, i);
        }

        vector<vector<int>> adj1(maxAdjSize + 1), adj2(maxAdjSize + 1);
        for (int i = 0; i < n - 1; i++)
        {
            for (int j = i + 1; j < n; j++)
            {
                if (rating[i] > rating[j])
                {
                    adj1[rating[i]].push_back(rating[j]);
                }
                else if (rating[i] < rating[j])
                {
                    adj2[rating[i]].push_back(rating[j]);
                }
            }
        }

        int res = 0;
        for (int &i : rating)
        {
            for (int &j : adj1[i])
            {
                res += adj1[j].size();
            }
            for (int &j : adj2[i])
            {
                res += adj2[j].size();
            }
        }

        return res;
    }
};